### PR TITLE
chore(vrl): improve performance of `op` expression

### PR DESCRIPTION
**Pull request by JeanMertz**
Similar to #12679, but for `op` expressions.

This special-cases `Err` and `Or` operations, allowing the others to use less indirection in the code. Again, I doubt it'll have any meaningful impact on soaks, but I consider it still a worthy improvement.

### CODE REVIEW

File: lib/vrl/compiler/src/expression/op.rs [hunk 1]
```diff
@@ -93,28 +93,37 @@ impl Expression for Op {
...
-            Sub => lhs?.try_sub(rhs()?),
-            Rem => lhs?.try_rem(rhs()?),
-            Or => lhs?.try_or(rhs),
-            And => match lhs? {
+            Mul => lhs.try_mul(rhs),
+            Div => lhs.try_div(rhs),
+            Add => lhs.try_add(rhs),
+            Sub => lhs.try_sub(rhs),
+            Rem => lhs.try_rem(rhs),
+            And => match lhs {
```
Comment (fuchsnj) [inline lib/vrl/compiler/src/expression/op.rs:115]: `And` also needs to short-circuit if the first value is false

**Review by fuchsnj — CHANGES_REQUESTED** [review]

File: lib/vrl/compiler/src/expression/op.rs (hunk 1 above)
Comment (JeanMertz) [reply lib/vrl/compiler/src/expression/op.rs:115]: Good catch!

File: lib/vrl/compiler/src/expression/op.rs [hunk 2]
```diff
@@ -93,28 +93,37 @@ impl Expression for Op {
...
+            return self.lhs.resolve(ctx).or_else(|_| self.rhs.resolve(ctx));
+        } else if let Or = self.opcode {
+            return self
+                .lhs
+                .resolve(ctx)?
+                .try_or(|| self.rhs.resolve(ctx))
+                .map_err(Into::into);
+        };
+
+        let lhs = self.lhs.resolve(ctx)?;
```
Comment (fuchsnj) [inline lib/vrl/compiler/src/expression/op.rs:106]: Why not move this to the top? `lhs` is _always_ resolved.

Comment (JeanMertz) [reply lib/vrl/compiler/src/expression/op.rs:111]: It's always resolved, but for `Err` it's not short-circuited (note the `?`).

File: lib/vrl/compiler/src/expression/op.rs (hunk 1 above)
Comment (JeanMertz) [reply lib/vrl/compiler/src/expression/op.rs]: Resolved in [60ca3a1](https://github.com/vectordotdev/vector/pull/12681/commits/60ca3a1de49851670fd063048b7ae4e21652d818).

**Review by fuchsnj — APPROVED** [review]

File: lib/vrl/compiler/src/expression/op.rs [hunk 3]
```diff
@@ -93,28 +93,38 @@ impl Expression for Op {
         use ast::Opcode::*;
         use Value::*;
 
-        let lhs = self.lhs.resolve(ctx);
-        let mut rhs = || self.rhs.resolve(ctx);
+        if let Err = self.opcode {
```
[the remark this answers is not in the document]
Comment (JeanMertz) [reply lib/vrl/compiler/src/expression/op.rs:96]: Added it to the list of "compiler improvements" I'll be working on in a bit 👍

**Pull request merged**